cocos/core/geometry/spline.ts

```diff
@@ -0,0 +1,292 @@
+     * The whole curve passes through all knots.
+     */
+    CATMULL_ROM = 2
+};
+
+const SPLINE_WHOLE_INDEX = 0xffffffff;
+
+const _v0 = new Vec3();
+const _v1 = new Vec3();
+const _v2 = new Vec3();
+const _v3 = new Vec3();
+
+/**
+ * @en
+ * Basic Geometry: Spline.
+ * @zh
+ * 基础几何 Spline
+ */
+
+export class Spline {
+    private readonly _type: number;
+    private _mode: SplineMode = SplineMode.CATMULL_ROM;
+    private _knots: Vec3[] = [];
+
+    public constructor(mode: SplineMode = SplineMode.CATMULL_ROM, knots: Vec3[] = []) {
+        this._type = enums.SHAPE_SPLINE;
+        this._mode = mode;
+
+        for (let i = 0; i < knots.length; i++) {
+            this._knots[i] = new Vec3(knots[i]);
+        }
+    }
+
+    public static create (mode: SplineMode, knots: Vec3[] = []) {
+        return new Spline(mode, knots);
+    }
+
+    public static clone (s: Spline) {
+        return new Spline(s.getMode(), s.getKnots());
+    }
+
+    public static copy (out: Spline, s: Spline) {
+        out._mode = s.getMode();
+        out._knots = [];
+
+        for (let i = 0; i < s.getKnots().length; i++) {
+            out._knots[i] = new Vec3(s.getKnots()[i]);
+        }
+
+        return out;
+    }
+
+    get type () {
+        return this._type;
+    }
+
+    public getMode () {
+        return this._mode;
+    }
+
+    public setMode (value) {
+        this._mode = value;
+    }
+
+    public getKnots () {
+        return this._knots;
+    }
+
+    public setKnots (value) {
+        this._knots = [];
+
+        for (let i = 0; i < value.length; i++) {
+            this._knots[i] = new Vec3(value[i]);
+        }
+    }
+
+    public clearKnots() { 
+        this._knots = [];
+    }
+
+    public getKnotCount() { 
+        return this._knots.length; 
+    }
+
+    public addKnot(knot: Vec3) { 
+        this._knots.push(new Vec3(knot)); 
+    }
+
+    public insertKnot(index: number, knot: Vec3) {
+        const item = new Vec3(knot);
+        if (index >= this._knots.length) {
+            this._knots.push(item);
+            return;
+        }
+    
+        this._knots.splice(index, 0, item);
+    }
+
+    public removeKnot(index: number) {
+        assertIsTrue(index < this._knots.length, "Spline: invalid index");
+
+        this._knots.splice(index, 1);
+    }
+
+    public setKnot(index: number, knot: Vec3) {
+        assertIsTrue(index < this._knots.length, "Spline: invalid index");
+
+        this._knots[index].set(knot);
+    }
+
+    public getKnot(index: number) {
+        assertIsTrue(index < this._knots.length, "Spline: invalid index");
+
+        return this._knots[index];
+    }
+
+    // get a point at t with repect to the `index` segment of curve or the whole curve.
+    public getPoint(t: number, index: number = SPLINE_WHOLE_INDEX) {
+        t = clamp(t, 0.0, 1.0);
+
+        const segments = this.getSegments();
+        if (segments == 0) {
+            return new Vec3(0.0, 0.0, 0.0);
```

Comment:
In TS, it is better to use global Vec3 defined in the file.

Reply:
This method allows the user to modify the return value